[AIRFLOW-4462] Fix incorrect datetime column types when using MSSQL backend

deodeveloper opened the pull request:
- Airflow manual trigger/scheduled trigger doesn't always work on mssql(azure mssql, mssql 2017, other versions)

* fixed mssql server as airflow backend. Applies only to sql server 2005  later versions and azure sql db where datetime2 datatype is available. Issue is with execution_date used as primary key in task_instance and a equality comparison column in dag_run. When comparing equality python microsecond precision to sql server, it returns false when the millisecond is non-zero,e.g, 2019-04-20T18:51:35.033. This change will make the execution_date from datetime to datetime(precision=6) matches with mysql precision. The date equality comparision works as desired. Microsoft also recommends to use datetime2.

Without this fix, airflow is not usable in sql server and azure sql server db. This fix can also be applied to existing stable branches to make airflow work with mssql server

Closed #5241 and created this pull request against master as per the suggestion from @potiuk 

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

Commit:
[AIRFLOW-4462] MSSQL backend broken - Airflow manual trigger/scheduled trigger doesn't always work on mssql(azure mssql, mssql 2017, other versions)
* fix mssql server as airflow backend. Applies only to sql server 2005  later versions and azure sql db where datetime2 datatype is available. Issue is with execution_date used as primary key in task_instance and a equality comparison column in dag_run. When comparing equality python microsecond precision to sql server, it returns false when the millisecond is non-zero,e.g, 2019-04-20T18:51:35.033. This change will make the execution_date from datetime to datetime(precision=6) matches with mysql precision. The date equality comparision works as desired. Microsoft also recommends to use datetime2.

Without this fix, airflow is not usable in sql server and azure sql server db. This fix can also be applied to existing stable branches to make airflow work with mssql server

ashb commented on airflow/migrations/versions/464861dc0797_change_execution_date_from_datetime_to_.py:
```diff
@@ -0,0 +1,294 @@
+        with op.batch_alter_table('task_reschedule') as task_reschedule_batch_op:
+            task_reschedule_batch_op.drop_index('idx_task_reschedule_dag_task_date')
+
+            task_reschedule_batch_op.drop_constraint(
```
Are we are dropping and re-creating this FK because otherwise we can't alter the column type?

deodeveloper replied:
@ashb  Yes, we are recreating this FK at line no. 79.
https://github.com/apache/airflow/pull/5255/files/09b22e97e1fdb4e7bdca0d31ccb766e8a2079ccb#diff-e887f8103d940732ec15b84a92f1dfdfR79

deodeveloper replied:
@ashb Yes, we are recreating this FK at line no. 79.
https://github.com/apache/airflow/pull/5255/files/09b22e97e1fdb4e7bdca0d31ccb766e8a2079ccb#diff-e887f8103d940732ec15b84a92f1dfdfR79

ashb reviewed (changes requested):
There are more columns that should store fractioinal seconds too. See https://github.com/apache/airflow/blob/master/airflow/migrations/versions/4addfa1236f1_add_fractional_seconds_to_mysql_tables.py for an (old) example of when we did this for MySql. (That list won't be complete/up-to-date with the current models, but there are more than just execution_date that we need to change I think)

Commit:
[AIRFLOW-4462]
added few more column data type changes to match with corresponding mysql changes mentioned here - https://github.com/apache/airflow/blob/master/airflow/migrations/versions/4addfa1236f1_add_fractional_seconds_to_mysql_tables.py
